File: app/src/main/java/com/example/myapplication/ui/screens/AuthScreen.kt
package com.example.myapplication.ui.screens

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.unit.dp

@Composable
fun AuthScreen(
    isLoading: Boolean,
    errorMessage: String?,
    infoMessage: String?,
    retryAfterSeconds: Int?,
    attemptsRemaining: Int?,
    showResendVerification: Boolean,
    admin2faEmail: String?,
    showOtpCodeField: Boolean,
    resendCooldownSeconds: Int?,
    forgotPasswordCooldownSeconds: Int?,
    onLogin: (email: String, password: String, otpCode: String?) -> Unit,
    onRegister: (email: String, password: String) -> Unit,
    onResendVerification: (email: String) -> Unit,
    onForgotPassword: (email: String) -> Unit
) {
    var isRegisterMode by remember { mutableStateOf(false) }
    var email by remember { mutableStateOf("") }
    var password by remember { mutableStateOf("") }
    var otpCode by remember { mutableStateOf("") }

    val isRateLimited = (retryAfterSeconds ?: 0) > 0
    val isAdminLoginTarget = admin2faEmail
        ?.takeIf { it.isNotBlank() }
        ?.equals(email.trim(), ignoreCase = true) == true
    val shouldShowOtpField = !isRegisterMode && (showOtpCodeField || isAdminLoginTarget)
    val hasMinLength = password.length >= 10
    val hasLower = password.any { it.isLowerCase() }
    val hasUpper = password.any { it.isUpperCase() }
    val hasDigit = password.any { it.isDigit() }
    val hasSymbol = password.any { !it.isLetterOrDigit() }
    val registerPasswordValid = hasMinLength && hasLower && hasUpper && hasDigit && hasSymbol
    val canSubmit = if (isRegisterMode) {
        email.contains("@") && registerPasswordValid && !isLoading && !isRateLimited
    } else {
        email.contains("@") &&
            password.isNotBlank() &&
            (!shouldShowOtpField || otpCode.length >= 6) &&
            !isLoading &&
            !isRateLimited
    }
    val resendCooldown = resendCooldownSeconds ?: 0
    val forgotCooldown = forgotPasswordCooldownSeconds ?: 0
    val canResendVerification = isRegisterMode &&
        showResendVerification &&
        email.contains("@") &&
        !isLoading &&
        !isRateLimited &&
        resendCooldown <= 0
    val canForgotPassword = !isRegisterMode &&
        email.contains("@") &&
        !isLoading &&
        !isRateLimited &&
        forgotCooldown <= 0
    val submitLabel = if (isRegisterMode) "Créer un compte" else "Se connecter"
    val modeTitle = if (isRegisterMode) "Inscription" else "Connexion"
    val modeSubtitle = if (isRegisterMode) {
        "Crée un compte personnel pour ta propre liste."
    } else {
        "Connecte-toi pour charger ta liste personnelle."
    }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .padding(horizontal = 24.dp),
        verticalArrangement = Arrangement.Center,
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Text(
            text = "WatchTracker",
            style = MaterialTheme.typography.headlineMedium
        )
        Text(
            text = modeTitle,
            style = MaterialTheme.typography.titleLarge,
            modifier = Modifier.padding(top = 8.dp)
        )
        Text(
            text = modeSubtitle,
            style = MaterialTheme.typography.bodyMedium,
            modifier = Modifier.padding(top = 8.dp, bottom = 24.dp)
        )

        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(bottom = 12.dp),
            horizontalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            if (!isRegisterMode) {
                Button(
                    onClick = { isRegisterMode = false },
                    enabled = !isLoading,
                    modifier = Modifier.weight(1f)
                ) {
                    Text("Connexion")
                }
            } else {
                OutlinedButton(
                    onClick = { isRegisterMode = false },
                    enabled = !isLoading,
                    modifier = Modifier.weight(1f)
                ) {
                    Text("Connexion")
                }
            }

            if (isRegisterMode) {
                Button(
                    onClick = { isRegisterMode = true },
                    enabled = !isLoading,
                    modifier = Modifier.weight(1f)
                ) {
                    Text("Inscription")
                }
            } else {
                OutlinedButton(
                    onClick = { isRegisterMode = true },
                    enabled = !isLoading,
                    modifier = Modifier.weight(1f)
                ) {
                    Text("Inscription")
                }
            }
        }

        OutlinedTextField(
            value = email,
            onValueChange = { email = it },
            label = { Text("Adresse mail") },
            enabled = !isLoading,
            singleLine = true,
            modifier = Modifier.fillMaxWidth()
        )

        OutlinedTextField(
            value = password,
            onValueChange = { password = it },
            label = {
                if (isRegisterMode) {
                    Text("Mot de passe (10+, min/maj/chiffre/symbole)")
                } else {
                    Text("Mot de passe")
                }
            },
            enabled = !isLoading,
            singleLine = true,
            visualTransformation = PasswordVisualTransformation(),
            modifier = Modifier
                .fillMaxWidth()
                .padding(top = 12.dp)
        )

        if (shouldShowOtpField) {
            OutlinedTextField(
                value = otpCode,
                onValueChange = { value ->
                    otpCode = value.filter { it.isDigit() }.take(8)
                },
                label = { Text("Code 2FA (admin)") },
                enabled = !isLoading,
                singleLine = true,
                keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.NumberPassword),
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(top = 12.dp)
            )
        }

        if (isRegisterMode) {
            val checks = listOf(
                "10 caractères minimum" to hasMinLength,
                "au moins une minuscule" to hasLower,
                "au moins une majuscule" to hasUpper,
                "au moins un chiffre" to hasDigit,
                "au moins un symbole" to hasSymbol
            )
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(top = 8.dp)
            ) {
                checks.forEach { (label, isValid) ->
                    val prefix = if (isValid) "• ✓ " else "• ✗ "
                    val color = if (isValid) {
                        Color(0xFF2E7D32)
                    } else {
                        MaterialTheme.colorScheme.error
                    }
                    Text(
                        text = prefix + label,
                        color = color,
                        style = MaterialTheme.typography.bodySmall,
                        modifier = Modifier.padding(top = 2.dp)
                    )
                }
            }
        }

        if (!errorMessage.isNullOrBlank()) {
            Text(
                text = errorMessage,
                color = MaterialTheme.colorScheme.error,
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(top = 12.dp)
            )
        }

        if (!infoMessage.isNullOrBlank()) {
            Text(
                text = infoMessage,
                color = MaterialTheme.colorScheme.primary,
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(top = 12.dp)
            )
        }

        if (isRateLimited) {
            Text(
                text = "Trop de tentatives. Réessaie dans ${retryAfterSeconds}s.",
                color = MaterialTheme.colorScheme.error,
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(top = 8.dp)
            )
        } else if (attemptsRemaining != null) {
            Text(
                text = "Tentatives restantes avant blocage : $attemptsRemaining",
                color = MaterialTheme.colorScheme.error,
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(top = 8.dp)
            )
        }

        Button(
            onClick = {
                val cleanEmail = email.trim()
                if (isRegisterMode) {
                    onRegister(cleanEmail, password)
                } else {
                    onLogin(cleanEmail, password, if (shouldShowOtpField) otpCode else null)
                }
            },
            enabled = canSubmit,
            modifier = Modifier
                .fillMaxWidth()
                .padding(top = 20.dp)
        ) {
            if (isLoading) {
                CircularProgressIndicator(
                    modifier = Modifier.padding(2.dp),
                    strokeWidth = 2.dp
                )
            } else {
                Text(submitLabel)
            }
        }

        if (!isRegisterMode) {
            OutlinedButton(
                onClick = { onForgotPassword(email.trim()) },
                enabled = canForgotPassword,
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(top = 10.dp)
            ) {
                if (forgotCooldown > 0) {
                    Text("Mot de passe oublié (${forgotCooldown}s)")
                } else {
                    Text("Mot de passe oublié")
                }
            }
        }

        if (isRegisterMode && showResendVerification) {
            OutlinedButton(
                onClick = { onResendVerification(email.trim()) },
                enabled = canResendVerification,
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(top = 10.dp)
            ) {
                if (resendCooldown > 0) {
                    Text("Renvoyer l'email (${resendCooldown}s)")
                } else {
                    Text("Renvoyer l'email de vérification")
                }
            }
        }

    }
}

File: flutter_app/lib/src/presentation/screens/auth_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

class AuthScreen extends StatefulWidget {
  const AuthScreen({
    super.key,
    required this.appVersionLabel,
    required this.isLoading,
    required this.onLogin,
    required this.onRegister,
    required this.onForgotPassword,
    required this.onResendVerification,
    this.errorMessage,
    this.infoMessage,
    this.retryAfterSeconds,
    this.attemptsRemaining,
    this.showResendVerification = false,
    this.showOtpCodeField = false,
    this.admin2faEmail,
    this.resendCooldownSeconds,
    this.forgotCooldownSeconds,
    this.onModeChanged,
  });

  final bool isLoading;
  final String appVersionLabel;
  final String? errorMessage;
  final String? infoMessage;
  final int? retryAfterSeconds;
  final int? attemptsRemaining;
  final bool showResendVerification;
  final bool showOtpCodeField;
  final String? admin2faEmail;
  final int? resendCooldownSeconds;
  final int? forgotCooldownSeconds;
  final Future<void> Function(String email, String password, String? otpCode)
  onLogin;
  final Future<void> Function(String email, String password) onRegister;
  final Future<void> Function(String email) onForgotPassword;
  final Future<void> Function(String email) onResendVerification;
  final VoidCallback? onModeChanged;

  @override
  State<AuthScreen> createState() => _AuthScreenState();
}

class _AuthScreenState extends State<AuthScreen> {
  final _emailCtrl = TextEditingController();
  final _passwordCtrl = TextEditingController();
  final _otpCtrl = TextEditingController();
  bool _isRegisterMode = false;
  bool _showPassword = false;

  bool get _hasMinLength => _passwordCtrl.text.length >= 10;
  bool get _hasLower => _passwordCtrl.text
      .split('')
      .any((c) => c.toLowerCase() != c.toUpperCase() && c == c.toLowerCase());
  bool get _hasUpper => _passwordCtrl.text
      .split('')
      .any((c) => c.toLowerCase() != c.toUpperCase() && c == c.toUpperCase());
  bool get _hasDigit =>
      _passwordCtrl.text.split('').any((c) => '0123456789'.contains(c));
  bool get _hasSymbol => _passwordCtrl.text
      .split('')
      .any((c) => !RegExp(r'[A-Za-z0-9]').hasMatch(c));
  bool get _passwordValid =>
      _hasMinLength && _hasLower && _hasUpper && _hasDigit && _hasSymbol;

  bool get _isAdminTarget =>
      widget.admin2faEmail != null &&
      _emailCtrl.text.trim().toLowerCase() ==
          widget.admin2faEmail!.toLowerCase();

  bool get _shouldShowOtp =>
      !_isRegisterMode && (widget.showOtpCodeField || _isAdminTarget);

  bool get _isRateLimited => (widget.retryAfterSeconds ?? 0) > 0;

  bool get _canSubmit {
    if (widget.isLoading || _isRateLimited) return false;
    if (!_emailCtrl.text.contains('@')) return false;
    if (_isRegisterMode) return _passwordValid;
    if (_passwordCtrl.text.isEmpty) return false;
    if (_shouldShowOtp && _otpCtrl.text.length < 6) return false;
    return true;
  }

  bool get _canResend =>
      widget.showResendVerification &&
      _emailCtrl.text.contains('@') &&
      !widget.isLoading &&
      !_isRateLimited &&
      (widget.resendCooldownSeconds ?? 0) <= 0;

  bool get _canForgot =>
      !_isRegisterMode &&
      _emailCtrl.text.contains('@') &&
      !widget.isLoading &&
      !_isRateLimited &&
      (widget.forgotCooldownSeconds ?? 0) <= 0;

  @override
  void dispose() {
    _emailCtrl.dispose();
    _passwordCtrl.dispose();
    _otpCtrl.dispose();
    super.dispose();
  }

  Future<void> _submit() async {
    final email = _emailCtrl.text.trim();
    final password = _passwordCtrl.text;
    try {
      if (_isRegisterMode) {
        await widget.onRegister(email, password);
      } else {
        await widget.onLogin(
          email,
          password,
          _shouldShowOtp ? _otpCtrl.text.trim() : null,
        );
      }
    } catch (_) {
      // errors are handled by parent via errorMessage prop
    }
  }

  @override
  Widget build(BuildContext context) {
    final resendCooldown = widget.resendCooldownSeconds ?? 0;
    final forgotCooldown = widget.forgotCooldownSeconds ?? 0;

    return Scaffold(
      resizeToAvoidBottomInset: false,
      body: SafeArea(
        child: Column(
          children: [
            Expanded(
              child: SingleChildScrollView(
                padding: const EdgeInsets.symmetric(horizontal: 24),
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.center,
                  children: [
                    const SizedBox(height: 48),
                    Image.asset(
                      'assets/images/watchtracker_logo.png',
                      width: 110,
                      height: 110,
                      errorBuilder: (_, _, _) => const Icon(
                        Icons.movie_filter_rounded,
                        size: 110,
                        color: Color(0xFF006A6A),
                      ),
                    ),
                    const SizedBox(height: 4),
                    Text(
                      'WatchTracker',
                      style: Theme.of(context).textTheme.headlineMedium,
                    ),
                    const SizedBox(height: 4),
                    Text(
                      _isRegisterMode ? 'Inscription' : 'Connexion',
                      style: Theme.of(context).textTheme.titleLarge,
                    ),
                    const SizedBox(height: 4),
                    Text(
                      _isRegisterMode
                          ? 'Crée un compte personnel pour ta propre liste.'
                          : 'Connecte-toi pour charger ta liste personnelle.',
                      style: Theme.of(context).textTheme.bodyMedium,
                      textAlign: TextAlign.center,
                    ),
                    const SizedBox(height: 24),
                    Row(
                      children: [
                        Expanded(
                          child: _isRegisterMode
                              ? OutlinedButton(
                                  onPressed: widget.isLoading
                                      ? null
                                      : () {
                                          setState(
                                            () => _isRegisterMode = false,
                                          );
                                          widget.onModeChanged?.call();
                                        },
                                  child: const Text('Connexion'),
                                )
                              : FilledButton(
                                  onPressed: null,
                                  child: const Text('Connexion'),
                                ),
                        ),
                        const SizedBox(width: 8),
                        Expanded(
                          child: _isRegisterMode
                              ? FilledButton(
                                  onPressed: null,
                                  child: const Text('Inscription'),
                                )
                              : OutlinedButton(
                                  onPressed: widget.isLoading
                                      ? null
                                      : () {
                                          setState(
                                            () => _isRegisterMode = true,
                                          );
                                          widget.onModeChanged?.call();
                                        },
                                  child: const Text('Inscription'),
                                ),
                        ),
                      ],
                    ),
                    const SizedBox(height: 12),
                    _OutlinedTextField(
                      controller: _emailCtrl,
                      label: 'Adresse mail',
                      keyboardType: TextInputType.emailAddress,
                      enabled: !widget.isLoading,
                      onChanged: (_) => setState(() {}),
                    ),
                    const SizedBox(height: 12),
                    _OutlinedTextField(
                      controller: _passwordCtrl,
                      label: _isRegisterMode
                          ? 'Mot de passe (10+, min/maj/chiffre/symbole)'
                          : 'Mot de passe',
                      obscureText: !_showPassword,
                      enabled: !widget.isLoading,
                      onChanged: (_) => setState(() {}),
                      suffixIcon: IconButton(
                        onPressed: () =>
                            setState(() => _showPassword = !_showPassword),
                        icon: Icon(
                          _showPassword
                              ? Icons.visibility_off_outlined
                              : Icons.visibility_outlined,
                        ),
                        tooltip: _showPassword
                            ? 'Masquer le mot de passe'
                            : 'Afficher le mot de passe',
                      ),
                    ),
                    if (_shouldShowOtp) ...[
                      const SizedBox(height: 12),
                      _OutlinedTextField(
                        controller: _otpCtrl,
                        label: 'Code 2FA (admin)',
                        keyboardType: TextInputType.number,
                        inputFormatters: [
                          FilteringTextInputFormatter.digitsOnly,
                          LengthLimitingTextInputFormatter(8),
                        ],
                        enabled: !widget.isLoading,
                        onChanged: (_) => setState(() {}),
                      ),
                    ],
                    if (_isRegisterMode) ...[
                      const SizedBox(height: 8),
                      SizedBox(
                        width: double.infinity,
                        child: _PasswordCriteria(
                          checks: [
                            ('10 caractères minimum', _hasMinLength),
                            ('au moins une minuscule', _hasLower),
                            ('au moins une majuscule', _hasUpper),
                            ('au moins un chiffre', _hasDigit),
                            ('au moins un symbole', _hasSymbol),
                          ],
                        ),
                      ),
                    ],
                    if (widget.errorMessage != null &&
                        widget.errorMessage!.isNotEmpty) ...[
                      const SizedBox(height: 12),
                      Align(
                        alignment: Alignment.centerLeft,
                        child: Text(
                          widget.errorMessage!,
                          style: TextStyle(
                            color: Theme.of(context).colorScheme.error,
                          ),
                        ),
                      ),
                    ],
                    if (widget.infoMessage != null &&
                        widget.infoMessage!.isNotEmpty) ...[
                      const SizedBox(height: 12),
                      Align(
                        alignment: Alignment.centerLeft,
                        child: Text(
                          widget.infoMessage!,
                          style: TextStyle(
                            color: Theme.of(context).colorScheme.primary,
                          ),
                        ),
                      ),
                    ],
                    if (_isRateLimited) ...[
                      const SizedBox(height: 8),
                      Align(
                        alignment: Alignment.centerLeft,
                        child: Text(
                          'Trop de tentatives. Réessaie dans ${widget.retryAfterSeconds}s.',
                          style: TextStyle(
                            color: Theme.of(context).colorScheme.error,
                          ),
                        ),
                      ),
                    ] else if (widget.attemptsRemaining != null) ...[
                      const SizedBox(height: 8),
                      Align(
                        alignment: Alignment.centerLeft,
                        child: Text(
                          'Tentatives restantes avant blocage : ${widget.attemptsRemaining}',
                          style: TextStyle(
                            color: Theme.of(context).colorScheme.error,
                          ),
                        ),
                      ),
                    ],
                    const SizedBox(height: 20),
                    SizedBox(
                      width: double.infinity,
                      child: FilledButton(
                        onPressed: _canSubmit ? _submit : null,
                        child: widget.isLoading
                            ? const SizedBox(
                                height: 20,
                                width: 20,
                                child: CircularProgressIndicator(
                                  strokeWidth: 2,
                                  color: Colors.white,
                                ),
                              )
                            : Text(
                                _isRegisterMode
                                    ? 'Créer un compte'
                                    : 'Se connecter',
                              ),
                      ),
                    ),
                    if (!_isRegisterMode) ...[
                      const SizedBox(height: 10),
                      SizedBox(
                        width: double.infinity,
                        child: OutlinedButton(
                          onPressed: _canForgot
                              ? () => widget.onForgotPassword(
                                  _emailCtrl.text.trim(),
                                )
                              : null,
                          child: Text(
                            forgotCooldown > 0
                                ? 'Mot de passe oublié (${forgotCooldown}s)'
                                : 'Mot de passe oublié',
                          ),
                        ),
                      ),
                    ],
                    if (_isRegisterMode && widget.showResendVerification) ...[
                      const SizedBox(height: 10),
                      SizedBox(
                        width: double.infinity,
                        child: OutlinedButton(
                          onPressed: _canResend
                              ? () => widget.onResendVerification(
                                  _emailCtrl.text.trim(),
                                )
                              : null,
                          child: Text(
                            resendCooldown > 0
                                ? 'Renvoyer l\'email (${resendCooldown}s)'
                                : 'Renvoyer l\'email de vérification',
                          ),
                        ),
                      ),
                    ],
                    const SizedBox(height: 20),
                  ],
                ),
              ),
            ),
            Padding(
              padding: const EdgeInsets.fromLTRB(16, 4, 16, 10),
              child: Text(
                'Version ${widget.appVersionLabel}',
                textAlign: TextAlign.center,
                style: Theme.of(context).textTheme.bodySmall?.copyWith(
                  color: Theme.of(context).colorScheme.onSurfaceVariant,
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}

class _OutlinedTextField extends StatelessWidget {
  const _OutlinedTextField({
    required this.controller,
    required this.label,
    this.obscureText = false,
    this.keyboardType,
    this.inputFormatters,
    this.enabled = true,
    this.onChanged,
    this.suffixIcon,
  });

  final TextEditingController controller;
  final String label;
  final bool obscureText;
  final TextInputType? keyboardType;
  final List<TextInputFormatter>? inputFormatters;
  final bool enabled;
  final ValueChanged<String>? onChanged;
  final Widget? suffixIcon;

  @override
  Widget build(BuildContext context) {
    return TextField(
      controller: controller,
      obscureText: obscureText,
      keyboardType: keyboardType,
      inputFormatters: inputFormatters,
      enabled: enabled,
      onChanged: onChanged,
      decoration: InputDecoration(
        labelText: label,
        border: const OutlineInputBorder(),
        suffixIcon: suffixIcon,
      ),
    );
  }
}

class _PasswordCriteria extends StatelessWidget {
  const _PasswordCriteria({required this.checks});

  final List<(String, bool)> checks;

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: checks.map((entry) {
        final (label, isValid) = entry;
        return Padding(
          padding: const EdgeInsets.only(top: 2),
          child: Text(
            '${isValid ? "• ✓ " : "• ✗ "}$label',
            style: Theme.of(context).textTheme.bodySmall?.copyWith(
              color: isValid
                  ? const Color(0xFF2E7D32)
                  : Theme.of(context).colorScheme.error,
            ),
          ),
        );
      }).toList(),
    );
  }
}
